superset-frontend/src/components/ErrorMessage/ParameterErrorMessage.tsx

```diff
@@ -32,17 +33,76 @@ interface ParameterErrorExtra {
   }[];
 }
 
-const triggerMessage = t('This may be triggered by:');
+const maxDistanceForSuggestion = 2;
+const findMatches = (
+  undefined_parameters: string[],
+  template_keys: string[],
+) => {
+  const matches: { [undefined_parameter: string]: string[] } = {};
+  undefined_parameters.forEach(undefined_parameter => {
+    template_keys.forEach(template_key => {
+      if (
+        levenshtein(undefined_parameter, template_key) <=
+        maxDistanceForSuggestion
+      ) {
+        if (!matches[undefined_parameter]) {
+          matches[undefined_parameter] = [];
+        }
+        matches[undefined_parameter].push(`"${template_key}"`);
+      }
+    });
+  });
+  return matches;
+};
 
 function ParameterErrorMessage({
   error,
   source = 'sqllab',
 }: ErrorMessageComponentProps<ParameterErrorExtra>) {
   const { extra, level, message } = error;
 
+  const triggerMessage = tn(
+    'This was triggered by:',
+    'This may be triggered by:',
+    extra.issue_codes.length,
+  );
+
+  const matches = findMatches(
+    extra.undefined_parameters || [],
+    Object.keys(extra.template_parameters || []),
+  );
+
   const body = (
     <>
       <p>
+        {matches && (
+          <>
+            <p>{t('Did you mean:')}</p>
+            <ul>
+              {Object.entries(matches).map(
+                ([undefined_parameter, template_keys]) => (
+                  <li>
+                    {tn(
+                      '%(suggestion)s instead of "%(undefined)s?"',
```

Comment:
is it possible to centralize all the messages as const at the top?

Reply:
In this case the `tn` function depends on the runtime to determine which string to use, so it has to be called here.

I think something like this could work:

```js
const s1 = t('foo');
const s2 = t('bar');
...
function someFunction() => {
  ...
  const message = tn(s1, s2, numberOfItems);
}
```

But then it might make the translation harder, since the strings are disassociated? I'm not sure how the translation tooling works on the frontend.

@ktmud any thoughts here?

Reply:
Translations will not be available until they were read from the bootstrap data via [`preamble.ts`](https://github.com/airbnb/incubator-superset/blob/9a59bdda486f90b8faeafdc79af3bcdbb7c3d250/superset-frontend%2Fsrc%2Fpreamble.ts#L32) (prepended to every JS entry point script).

I think the usage here is fine, but I would recommend rename `undefined` to `undefined_vars` (including on the Python side), as `undefined` is a reserved word in JavaScript. This may cause confusion.

Another solution is to instead of using translation strings, you could present them as a two column table with "Variable Used" and "Suggestion(s)".